app/services/authentication/providers/apple.rb

```diff
@@ -0,0 +1,63 @@
+module Authentication
+  module Providers
+    # Apple authentication provider, uses omniauth-apple as backend
+    class Apple < Provider
+      SETTINGS_URL = "https://appleid.apple.com/account/manage".freeze
+
+      def new_user_data
+        # Apple sends `first_name` and `last_name` as separate fields
+        name = "#{info.first_name} #{info.last_name}"
+        timestamp = raw_info.auth_time || raw_info.id_info.auth_time
+
+        {
+          email: info.email,
+          apple_created_at: Time.zone.at(timestamp),
+          apple_username: user_nickname,
+          name: name,
+          remote_profile_image_url: SiteConfig.mascot_image_url
+        }
+      end
+
+      def existing_user_data
+        # Apple by default will send nil `first_name` and `last_name` after
+        # the first login. To cover the case where a user disconnects their
+        # Apple authorization, signs in again and then changes their name,
+        # we update the username only if the name is not nil
+        apple_username = info.first_name.present? ? info.first_name.downcase : nil
+        timestamp = raw_info.auth_time || raw_info.id_info.auth_time
+
+        data = { apple_created_at: Time.zone.at(timestamp) }
+        data[:apple_username] = apple_username if apple_username
+        data
+      end
+
+      # For Apple we override this method because the `info` payload doesn't
+      # include `nickname`. On top of not having a username, Apple allows users
+      # to 'choose' the first_name & last_name sent our way so they are
+      # definitely not assured to be unique. We still need `user_nickname` to
+      # always be the same on each login so we use the email hash as suffix to
+      # avoid collisions with other registrations with the same first_name
+      def user_nickname
+        email_digest = Digest::SHA512.hexdigest(info.email)
+        "#{info.first_name&.downcase}_#{email_digest}"[0...25]
+      end
```

Comment:
I left the note explaining why the need for the hash but it comes down to multiple users coming in with the same name. This is overriding `user_nickname` from `Provider.rb` so it remains clean/simple in the Base class and customizable for Providers that require some extra logic. (Anyone) feel free to ask/discuss

In this case the username is based on the user's "real name" which makes it seem a bit more natural than just a hash and looks like this:

<img width="699" alt="Screen Shot 2020-08-06 at 16 58 36" src="https://user-images.githubusercontent.com/6045239/89611397-58ef4500-d83a-11ea-8eb7-31562db5a512.png">

Reply:
I did implement something like this initially and then reverted it as we have an automatic mechanism who handle's collision. See https://github.com/forem/forem/pull/7952/commits/b82e5c71c380f4ccf79436035633ce4ee2f47516 which is a revert due to this comment https://github.com/forem/forem/pull/7952#issuecomment-643022179

Reply:
Absolutely! I actually saw what you did there and modified it a bit. The difference was that before their username would only be a hash (very cryptic username as in not personalized at all) and here it's their first name with the appended hash.

The automatic mechanism is made up by [these functions in User](https://github.com/fdoxyz/dev.to/blob/rhymes/add-sign-in-with-apple/app/models/user.rb#L510-L534) right? There is a problem with it though, the registration process will encounter an error ([in this save!](https://github.com/fdoxyz/dev.to/blob/rhymes/add-sign-in-with-apple/app/services/authentication/authenticator.rb#L47)) before making it to where the User is created (`before_validation :set_username` within User). The reason is that multiple Identities can be created that render the same "apple_username" and that's why I added this again. 

This is what I found happens:

1. User `johndoe@dev.to` signs up with `first_name="John" last_name="Doe"` and is assigned username `johndoe`
2. User `martin@othercompany.com` signs up but modifies their first & last name and also comes in as "johndoe".  [Authentication::Authenticator#find_or_create_user](https://github.com/fdoxyz/dev.to/blob/master/app/services/authentication/authenticator.rb#L98) checks and finds that a user with the username already exists, and instead of creating a new user it returns it and then an exception is raised because of this (before the automatic mechanism is processed)

The problem is not only that Apple OAuth process (when registering) users can change the name, but some common names will almost certainly collide. In the end `"#{first_name}_#{last_name}` is not guaranteed to be a unique way to know if a user is new or existing specifically within the Apple Provider "_namespace_".

The alternative I considered was not perfect either, and it consists of modifying `Authentication::Authenticator#find_or_create_user` so instead of this:

```
def find_or_create_user!
  existing_user = User.where(
    provider.user_username_field => provider.user_nickname,
  ).take
  return existing_user if existing_user
...
```

change it to look something like this:

```
def find_or_create_user!
  existing_user = User.where(
    provider.user_username_field => provider.user_nickname,
    'email' => provider.info.email,
  ).take
  return existing_user if existing_user
...
```

This way at least we know Apple won't send us users with the same username (regardless of how we define Apple users' username). __The problem I see with this__ is that common names will almost certainly repeat (a lot) and the existing mechanism uses a `rand(100)` call, so if we end up having 100 users with the same `first_name` & `last_name` (keep in mind people can "edit their name") it's likely for users to try to register and having to do lots of failed attempts until they get assigned a unique random username (once we have 100 it will become impossible for them to register).

It's all very complicated, if you sign up for Medium using Apple they use a worse username actually (doesn't mean we can't come up with a better solution!)

<img width="404" alt="Screen Shot 2020-08-07 at 09 43 21" src="https://user-images.githubusercontent.com/6045239/89663270-91227200-d892-11ea-9f49-ccc259d2c8c7.png">

I think an improvement (I can push this change soon) is to make this method look this way:

```
def user_nickname
  [
    info.first_name&.downcase,
    info.last_name&.downcase,
    Digest::SHA512.hexdigest(info.email),
  ].join("_")[0...25]
end
```

With this change the username would probably be something like `fernando_valverde_3a5f38a` instead of `fernando_3a5f38ad4e722de8`

Sorry for the super lengthy response! 😅